### ISSUE TITLE
Init not being run pre-launch

### ISSUE THREAD

**Issue by nhorton** [gitpod-io/gitpod#15187]
### Bug description

We have prebuilds in use heavily and are getting two seemingly related issues:
1. If we open a workspace straight from the branch URL, it does not use the prebuild
2. When launching without a prebuild, and *sometimes with a prebuild*, it runs the init again, in parallel with the command. That is obviously a big problem in many cases.

### Steps to reproduce

1. Turn on prebuilds
2. Let prebuilds complete.
3. Launch a workspace using the branch link added by the Badge option or the Github Actions PR section.

Note that the example Workspace below is one of the instances where it used a prebuild but still ran init

### Workspace affected

https://unsupervise-unsupervise-o8mkgiqoblm.ws-us77.gitpod.io/

### Expected behavior

Init should only run once, and prebuilds should always get used.

### Anything else?

We depend pretty heavily on prebuilds. It would be nice to have an option to force that prebuilds be used and never launched with Init done live.

Also - the linking of home directories into Workspace and similar for better persistence really has got to get baked in.

**Comment by nhorton** [comment]
I just got this again on https://unsupervise-unsupervise-9w2degsxtkg.ws-us78.gitpod.io/.

Really telling steps of what I did:
1. Went to the Prebuilds page in Gitpod this morning
2. Clicked on the green "Ready" for the latest prebuild that happened late last night.
3. Clicked "New Workspace (with this prebuild)" at the bottom-right
4. Workspace launched in ~30 seconds.
5. Had two Terminals open in VSCode - one with my Init still running (i.e. I literally see the progress indicators in the script coming through) and another with the Command output.

**Comment by svenefftinge** [comment]
Just to clarify:

The `init` task is meant to be idempotent, so it should be able to run multiple times on the same workspace. That we skip it on the workspace where on the exact same commit had a prebuild is an optimization. For instance, if you start a workspace using an older prebuild (ie.e. one that was done on a previous commit) Gitpod would run fetch and reset the git to the commit (or head of branch) you opened the workspace for and run `init` again.

Also, if you have multiple tasks Gitpod does not sync between them and run all `init` tasks before starting the `command` tasks. It only does this per task. If there are dependencies between the various commands you can use `gp sync` to sync between them.

Does that help with your situation?

**Comment by nhorton** [comment]
I do appreciate the info. The bit about multiple tasks has been referenced before but I have not actually seen a real doc on it. Specifically, I have a single "init" yml entry, but it is a series of steps in the shell. I presume that you mean if there were mutiple different init entiees, those would run in parallel, not the steps inside one init entry, right?

On the other parts idempotent is relative. The end-state is idempotent, but the whole idea of prebuilds is that stuff can take a while to run. Ex. I have a docker-compose pull in our init, so you can do that as often as you like. But prebuilds sort of lose their value if one of my employees launches a new workspace and it does not work for 15 minutes because it is redoing a long init sequence that was already done in the prebuild.

Directly related to that last task, our Command entry can't be run in the middle of the init and work.

**Comment by nhorton** [comment]
Also - if you want to screenshare sometime, I am happy to do it.

**Comment by svenefftinge** [comment]
The doc on multiple tasks is and that you can sync them is here https://www.gitpod.io/docs/configure/workspaces/tasks#wait-for-commands-to-complete

>  I presume that you mean if there were mutiple different init entiees, those would run in parallel, not the steps inside one init entry, right?

Every dash under tasks denotes a terminal and the steps in each terminal run sequentially, but we do not sync the lifecycle between the terminals, that you need to do manually using `gp sync`.

> On the other parts idempotent is relative. The end-state is idempotent, but the whole idea of prebuilds is that stuff can take a while to run.

The underlying idea is that prebuilds make use of any caches using the file system, as this is the typical caching mechanism for build systems. If the git repo hasn't changed we don't need to rerun any build steps, but if there is an even small change, it would be the responsibility of the build system to invalidate the right caches and run an incremental build. This is the assumption and why Gitpod re-runs the `init` task when you open a workspace using an older rebuild (i.e. one that wasn't executed on the exact commit you are on).

> Ex. I have a docker-compose pull in our init, so you can do that as often as you like. 

Docker usually does cache the pulled images. I wonder why they are pulled again in your case. Maybe you are really not starting workspaces on old prebuilds but without one. 

> Directly related to that last task, our Command entry can't be run in the middle of the init and work.

Sound like your gitpod.yml looks like this:
```
tasks:
 - init: do init
 - command: run command
```
So it is configured to run in two separate terminals. If you want to run one ofter the other just remove the second dash:
```
tasks:
 - init: do init
   command: run command
```
